dev/TEAL.md

```diff
@@ -147,8 +147,19 @@ For three-argument ops, `A` is the element two below the top, `B` is the penulti
 | `getbyte` | pop a byte-array A and integer B. Extract the Bth byte of A and push it as an integer |
 | `setbyte` | pop a byte-array A, integer B, and small integer C (between 0..255). Set the Bth byte of A to C, and push the result |
 | `concat` | pop two byte-arrays A and B and join them, push the result |
+
+These opcodes return portions of byte arrays, accessed by position, in
+various sizes.
+
+| Op | Description |
+| --- | --- |
 | `substring s e` | pop a byte-array A. For immediate values in 0..255 S and E: extract a range of bytes from A starting at S up to but not including E, push the substring result. If E < S, or either is larger than the array length, the program fails |
 | `substring3` | pop a byte-array A and two integers B and C. Extract a range of bytes from A starting at B up to but not including C, push the substring result. If C < B, or either is larger than the array length, the program fails |
+| `extract s l` | pop a byte-array A. For immediate values in 0..255 S and L: extract a range of bytes from A starting at S up to but not including S+L, push the substring result. If L is 0, then extract to the end of the string. If S or S+L is larger than the array length, the program fails |
+| `extract3` | pop a byte-array A and two integers B and C. Extract a range of bytes from A starting at B up to but not including B+C, push the substring result. If B or B+C is larger than the array length, the program fails |
```

Comment:
No, the thinking is that with `extract3` you have calculated the position, so 0 potentially "makes sense".  But with `extract`, 0 is an immediate value, so it would be nonsensical to use that instruction -- it just means `dup`.

Reply:
I'm a bit confused: in both cases you may want to remove the prefix.
It is possible that the length of the prefix is unknown.
An example of that would be when using the ABI where you encode `([]byte, addr)` for example.
You can always manually compute the remaining length, but why not allow to specify the length 0 for all the string?
The main drawback I see is that then when you use the instruction you may use it for length=0 for really length=0...

Reply:
That final sentence is my point.  extract3, you may have computed the length, and that computation might yield 0, and you may really mean it.

But it's nonsensical to use 0 as the immediate value with extract.  You are making a hardcoded choice to use 0.  That just yields the empty array. (it's not `dup` as I said above)

Reply:
That makes sense. I agree.

I would recommend that the assembler takes a special token `end` if possible instead of 0 in that case.
This way people reading/writing the code would not get confused.
The actual encoding of `end` would be 0, but the assembler would not accept `0` there, and only accept `end`.

i.e., you write `extract 10 end` instead of `extract 10 0`

Reply:
I like that.  Not sure it's doable right away, but it's nice.